File: agent/bank.go
package agent

import (
	"economy/company"
	"economy/market"
)

// Bank models a bank's own trading desk: conviction-driven and
// fundamentals-aware like HedgeFund, but distinguished by two things
// that are genuinely true of real bank desks rather than just "a fund
// with different numbers":
//
//  1. It trades multiple companies from one shared capital pool,
//     rather than one company per fund instance — real desks allocate
//     one balance sheet across many positions simultaneously.
//
//  2. It is subject to account-wide, drawdown-driven de-risking that
//     is independent of any individual position's thesis: as the
//     account's equity falls from its high-water mark, every
//     covered company's target exposure gets scaled down together.
//     This mirrors how a real risk committee or regulatory capital
//     constraint forces a desk to cut exposure across the board
//     during a bad stretch, rather than asking "is this specific
//     trade still a good idea" position by position. The book-wide
//     leverage cap itself (via market.Account.MaxLeverage) plays the
//     role of the hard regulatory capital floor; DrawdownDeRisk*
//     below plays the role of the more dynamic internal risk-limit
//     layer that sits alongside it in a real bank.
type Bank struct {
	id       string
	account  *market.Account
	coverage []*company.Company

	MinTradeThreshold       float64
	FullConvictionThreshold float64
	RebalanceThreshold      float64

	// SkepticismDiscount (0-1) — see convictionSizedOrder's
	// documentation and HedgeFund's field of the same name. Bank
	// defaults to a higher discount than HedgeFund, reflecting a bank
	// desk's typically more conservative, more heavily-scrutinized
	// research process relative to a hedge fund's.
	SkepticismDiscount float64

	// equityHighWaterMark tracks the highest account equity observed
	// so far, updated every tick this Bank is asked for orders.
	// Drawdown is measured relative to this peak, not to starting
	// capital, since a desk that's already given back some of an
	// earlier gain is genuinely in a worse risk position than one
	// sitting flat at its starting capital.
	equityHighWaterMark float64
	hwmInitialized      bool

	// DrawdownDeRiskThreshold is the fraction of drawdown from the
	// high-water mark (0.10 = 10% down from peak) at which risk
	// scaling begins to bite.
	DrawdownDeRiskThreshold float64

	// DrawdownFullCutoff is the drawdown fraction at which risk
	// scaling reaches zero — the desk stops adding any new or
	// incremental risk entirely (existing positions can still be
	// reduced, since a scale of 0 still allows delta to call for
	// unwinding toward a target of zero).
	DrawdownFullCutoff float64

	// externalPrices, when set via SetExternalPrices, is preferred
	// over currentPricesAcrossCoverage's ReportedValue-based fallback
	// for any covered symbol it contains.
	externalPrices map[string]float64
}

// NewBank constructs a Bank covering the given companies from one
// shared account. Threshold defaults are set more conservative than
// HedgeFund's — higher MinTradeThreshold (won't chase small
// mispricings the way an aggressive fund will) and lower per-company
// deployment, reflecting a real bank desk's more conservative mandate
// relative to a hedge fund's.
func NewBank(id string, acct *market.Account, coverage []*company.Company) *Bank {
	return &Bank{
		id:                      id,
		account:                 acct,
		coverage:                coverage,
		MinTradeThreshold:       0.06, // 6% — more conservative than HedgeFund's 3%
		FullConvictionThreshold: 0.30, // slightly higher bar for "fully convinced"
		RebalanceThreshold:      0.05,
		SkepticismDiscount:      0.50, // more conservative than HedgeFund's 0.35
		DrawdownDeRiskThreshold: 0.10, // de-risking begins at 10% drawdown from peak
		DrawdownFullCutoff:      0.30, // no new risk at all by 30% drawdown from peak
	}
}

func (b *Bank) ID() string               { return b.id }
func (b *Bank) Account() *market.Account { return b.account }

// riskScale computes the current account-wide risk-scaling factor
// from drawdown relative to the tracked high-water mark: 1.0 at no
// drawdown, linearly shrinking to 0 between DrawdownDeRiskThreshold
// and DrawdownFullCutoff, and 0 beyond DrawdownFullCutoff.
// currentEquity must reflect mark-to-market equity across everything
// this account holds, not just one company's position, since the
// de-risking response is meant to be account-wide.
func (b *Bank) riskScale(currentEquity float64) float64 {
	if !b.hwmInitialized || currentEquity > b.equityHighWaterMark {
		b.equityHighWaterMark = currentEquity
		b.hwmInitialized = true
		return 1.0
	}
	if b.equityHighWaterMark <= 0 {
		return 1.0 // avoid divide-by-zero in a degenerate account state
	}

	drawdown := (b.equityHighWaterMark - currentEquity) / b.equityHighWaterMark

	if drawdown <= b.DrawdownDeRiskThreshold {
		return 1.0
	}
	if drawdown >= b.DrawdownFullCutoff {
		return 0.0
	}
	span := b.DrawdownFullCutoff - b.DrawdownDeRiskThreshold
	return 1.0 - (drawdown-b.DrawdownDeRiskThreshold)/span
}

// currentPricesAcrossCoverage builds a symbol->price map covering
// every company this Bank trades. For the symbol the current tick's
// MarketState concerns, it uses the real market price (state.Mid).
// For every other covered symbol, it falls back to that company's own
// ReportedValue as a stand-in — consistent with this Bank reading
// ReportedValue elsewhere (see convictionSizedOrder), rather than
// TrueValue, which no participant in this simulation should read
// directly. This fallback is still an approximation: a real desk's
// mark-to-market equity needs the actual *market* price for every
// held symbol, not a fundamentals-based stand-in, for any symbol
// other than the one in the current tick's state. SetExternalPrices
// lets an orchestrator holding every company's real order book supply
// exact prices instead; when set, those take priority over this
// fallback entirely.
func (b *Bank) currentPricesAcrossCoverage(state market.MarketState) map[string]float64 {
	prices := make(map[string]float64, len(b.coverage))
	for _, c := range b.coverage {
		if b.externalPrices != nil {
			if p, ok := b.externalPrices[c.Symbol]; ok {
				prices[c.Symbol] = p
				continue
			}
		}
		if c.Symbol == state.Symbol && state.HasMid {
			prices[c.Symbol] = state.Mid
		} else {
			prices[c.Symbol] = c.ReportedValue
		}
	}
	return prices
}

// SetExternalPrices supplies a real, orchestrator-maintained
// symbol->market-price map that currentPricesAcrossCoverage should
// prefer over its ReportedValue-based fallback. An orchestrator
// holding every company's actual order book should call this once per
// tick before polling this Bank for orders, so cross-company
// mark-to-market equity (and therefore drawdown-driven de-risking) is
// computed from real market prices rather than an approximation. Safe
// to leave unset in isolated tests, where the fallback remains
// available.
func (b *Bank) SetExternalPrices(prices map[string]float64) {
	b.externalPrices = prices
}

// NextOrders implements market.OrderSource for a single covered
// company per call, matching how the sim loop is expected to drive
// any OrderSource: once per company's MarketState each tick. Bank
// identifies which of its covered companies state.Symbol refers to,
// updates its account-wide risk scale from current equity, and
// produces at most one conviction-sized order for that company scaled
// by the current risk factor.
func (b *Bank) NextOrders(state market.MarketState) []*market.Order {
	var target *company.Company
	for _, c := range b.coverage {
		if c.Symbol == state.Symbol {
			target = c
			break
		}
	}
	if target == nil {
		return nil // this Bank doesn't cover the company this tick's state concerns
	}

	equity := b.account.Equity(b.currentPricesAcrossCoverage(state))
	scale := b.riskScale(equity)

	order := convictionSizedOrder(
		b.id, b.account, target, state,
		b.MinTradeThreshold, b.FullConvictionThreshold, b.RebalanceThreshold,
		scale,
		b.SkepticismDiscount,
	)
	if order == nil {
		return nil
	}
	return []*market.Order{order}
}

File: src/agent/bank.rs
use std::collections::HashMap;
use std::sync::Arc;
use parking_lot::RwLock;

use crate::company::Company;
use crate::market::{Account, MarketState, Order, OrderSource};
use super::hedgefund::conviction_sized_order;

pub struct Bank {
    pub id: String,
    pub account: Arc<RwLock<Account>>,
    pub coverage: Vec<Company>,
    pub min_trade_threshold: f64,
    pub full_conviction_threshold: f64,
    pub rebalance_threshold: f64,
    pub skepticism_discount: f64,
    pub equity_high_water_mark: f64,
    pub hwm_initialized: bool,
    pub drawdown_de_risk_threshold: f64,
    pub drawdown_full_cutoff: f64,
    pub external_prices: Option<HashMap<String, f64>>,
    pub execution_rate: f64,
    pub max_order_shares: f64,
}

impl Bank {
    pub fn new(id: impl Into<String>, account: Arc<RwLock<Account>>, coverage: Vec<Company>) -> Self {
        Self {
            id: id.into(),
            account,
            coverage,
            min_trade_threshold: 0.06,
            full_conviction_threshold: 0.30,
            rebalance_threshold: 0.05,
            skepticism_discount: 0.50,
            equity_high_water_mark: 0.0,
            hwm_initialized: false,
            drawdown_de_risk_threshold: 0.10,
            drawdown_full_cutoff: 0.30,
            external_prices: None,
            execution_rate: 1.0,
            max_order_shares: 0.0,
        }
    }

    pub fn set_external_prices(&mut self, prices: HashMap<String, f64>) {
        self.external_prices = Some(prices);
    }

    pub fn risk_scale(&mut self, current_equity: f64) -> f64 {
        if !self.hwm_initialized || current_equity > self.equity_high_water_mark {
            self.equity_high_water_mark = current_equity;
            self.hwm_initialized = true;
            return 1.0;
        }
        if self.equity_high_water_mark <= 0.0 {
            return 1.0;
        }

        let drawdown = (self.equity_high_water_mark - current_equity) / self.equity_high_water_mark;
        if drawdown <= self.drawdown_de_risk_threshold {
            return 1.0;
        }
        if drawdown >= self.drawdown_full_cutoff {
            return 0.0;
        }
        let span = self.drawdown_full_cutoff - self.drawdown_de_risk_threshold;
        1.0 - (drawdown - self.drawdown_de_risk_threshold) / span
    }

    fn current_prices_across_coverage(&self, state: &MarketState) -> HashMap<String, f64> {
        let mut prices = HashMap::with_capacity(self.coverage.len());
        for c in &self.coverage {
            if let Some(ref ext) = self.external_prices {
                if let Some(&p) = ext.get(&c.symbol) {
                    prices.insert(c.symbol.clone(), p);
                    continue;
                }
            }
            if c.symbol == state.symbol && state.mid.is_some() {
                prices.insert(c.symbol.clone(), state.mid.unwrap());
            } else {
                prices.insert(c.symbol.clone(), c.reported_value);
            }
        }
        prices
    }
}

impl OrderSource for Bank {
    fn id(&self) -> &str {
        &self.id
    }

    fn next_orders(&mut self, state: &MarketState) -> Vec<Order> {
        let target = match self.coverage.iter().find(|c| c.symbol == state.symbol) {
            Some(t) => t.clone(),
            None => return Vec::new(),
        };

        let prices = self.current_prices_across_coverage(state);
        let equity = {
            let acct = self.account.read();
            acct.equity(&prices)
        };
        let scale = self.risk_scale(equity);

        let acct = self.account.read();
        let mut order = match conviction_sized_order(
            &self.id,
            &acct,
            &target,
            state,
            self.min_trade_threshold,
            self.full_conviction_threshold,
            self.rebalance_threshold,
            scale,
            self.skepticism_discount,
        ) {
            Some(o) => o,
            None => return Vec::new(),
        };

        if self.execution_rate > 0.0 && self.execution_rate < 1.0 {
            order.quantity *= self.execution_rate;
        }
        if self.max_order_shares > 0.0 && order.quantity > self.max_order_shares {
            order.quantity = self.max_order_shares;
        }
        if order.quantity <= 0.0 {
            return Vec::new();
        }

        vec![order]
    }
}
